### ISSUE TITLE
Builtin Zed prettier does not respect zed "tab_width" options

### ISSUE THREAD

**Issue by ikelos** [zed-industries/zed#20235]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

The `tab_size` option is set to 4 by default  (both globally and for javascript in particular) and the autoformatting on save is on by default.  Open a Javascript file, ensure the indents are 4 spaces, save the file.  The indents are changed to a tab size of 2.  This is because the in-built prettier (allowed by default) has its own `tab_width` defaults, which are set to 2.  This can be overridden in the zed config with:

```
{
    "languages": {
        "JavaScript": {
            "prettier": {
                "tab_width": 4
            }
        }
    }
}

```

However, this is extremely confusing and most users would expect `tab_size` at the top level, or at the language level to override the built-in prettier configuration.

This may be related to #12154 and #20009.

### Environment

Zed: v0.159.7 (Zed)
OS: Linux Wayland gentoo 2.15
Memory: 15.3 GiB
Architecture: x86_64
GPU: Intel(R) UHD Graphics (TGL GT1) || Intel open-source Mesa driver || Mesa 24.2.6

### If applicable, add mockups / screenshots to help explain present your vision of the feature

In my view, javascript files, when saved by default, should adhere to the tab_width defaults of zed.

### If applicable, attach your Zed.log file to this issue.

Not applicable, although when saving a broken JSON settings file, I discovered the following line:
```
2024-11-05T11:11:12.87589183Z [ERROR] default prettier instance failed to format buffer

Caused by:
    error during message '{"jsonrpc":"2.0","id":2,"method":"prettier/format","params":{"text":"..snip..","options":{"plugins":[],"parser":null,"filepath":"/home/mike/.config/zed/settings.json","prettierOptions":{"useTabs":false,"tabWidth":2,"printWidth":80}}}}' handling: SyntaxError: Unexpected token, expected "," (13:3)
      11 |   "buffer_font_size": 16,
      12 |   "tab_size": 4, "hard_tabs": false
    > 13 |   "theme": {
         |   ^
      14 |     "mode": "system",
      15 |     "light": "One Light",
      16 |     "dark": "One Dark"
```

Which caused me to investigate `prettierOptions` and eventually find that `prettier` has its own `tab_width` setting.

**Comment by notpeter** [comment]
I don't believe `tab_width` is a valid prettier setting. I believe it should be `tabWidth` but I would recommend against specifying these values for prettier in your Zed settings.  Prettier defaults to [`tabWidth = 2`](https://prettier.io/docs/options#tab-width), if you need to use a value of 4, you can set this value as:

```json
  "languages": {
    "JavaScript": {
      "tab_size": 4
    }
  }
```

Zed will inject this into the prettier tabWidth setting here:
https://github.com/zed-industries/zed/blob/8b3d315e406eb121b7b0030a7f39a4c9777a2e08/crates/prettier/src/prettier.rs#L372-L379

I also recommend you explicit include [`.prettierc.json`](https://prettier.io/docs/configuration) with similar prettier overrides per extension/language.  This way prettier on the CLI and in Zed will behave the same.

I understand your confusion as to why a global setting of `tab_size` does not override the per-language values, this is because `tab_size`, `hard_tabs`, etc may be overriden by the config.toml for languages:
https://github.com/zed-industries/zed/blob/8b3d315e406eb121b7b0030a7f39a4c9777a2e08/crates/languages/src/javascript/config.toml#L19
https://github.com/zed-industries/zed/blob/8b3d315e406eb121b7b0030a7f39a4c9777a2e08/crates/languages/src/go/config.toml#L16

The Zed defaults attempt to align with the most common defaults for a given language (e.g. hard tabs for go matching `go fmt`, `tab_size = 2` for matching prettier defaults for JS/TS/json/jsonc, etc.)

Since the default.json `tab_size` is 4, the Javascript config.toml overrides that default to 2.

**Comment by Pumacks** [comment]
Nahh this issue shouldnt be closed as **not planned**.
This is a valid feature.

I got Zed as a binary through pacman, this means i have to build this myself and have to edit these files for just adjusting the tabsize of the formatter. 
plus i cant update zed through my package manager now.

please reopen this issue.

**Comment by notpeter** [comment]
You shouldn't need to build yourself, you should just be able to add the following to your Zed  user settings as I suggested above:
```json
  "languages": {
    "JavaScript": {
      "tab_size": 4
    }
  }
```
Is this not working for you?

**Comment by ikelos** [comment]
I haven't tried it since I filed the bug, but certainly when I filed the bug that value was *ignored* because it was overwritten by the default pretier config (hence the workaround presented in the original bug).

**Comment by Pumacks** [comment]
> [@Pumacks](https://github.com/Pumacks) You shouldn't need to build yourself, you should just be able to add the following to your Zed settings (user settings or project settings) as I suggested above:
> 
>   "languages": {
>     "JavaScript": {
>       "tab_size": 4
>     }
>   }
> 
> Is this not working for you?

unfortunately not. :/

**Comment by notpeter** [comment]
I just double checked and those settings work correctly for me.

1. Note: You may need to duplicate this for multiple "languages" (`Typescript`, `TSX`, `JSX`) in addition to `Javascript` if you are editing file names other than `*.js`.
2. Do you have any settings which might conflict with the defaults? E.g. in your settings a top-level `formatter` key or `formatter` defined under the `languages.JavaScript` so you're directly invoking prettier with custom parameters or formatting via the language server?
3. Do you have a `.zed/settings.json` inside your project? (are there any relevant settings keys there).
4. Do you have an `.editorconfig` in your project? Or in any of the parent directories of the file you're working on? Because if an `.editorconfig` file is found these settings will in parent directories apply (e.g. `~/.editorconfig`)

---

Steps I followed with an a minimal zed settings.json:
```
cp ~/.config/zed/settings.json ~/.config/zed/settings.json.backup 
echo '{"languages": {"JavaScript": {"tab_size": 4}}}' > ~/.config/zed/settings.json

mkdir /tmp/issue20235
cd $_
echo 'function helloWorld() {console.log("Hello World!");}helloWorld();' > foo.js
zed . foo.js
```

Then Trigger "editor: format" from the command palette.
Does it still indent with 2 spaces? (it indents with 4 for me).

**Comment by Pumacks** [comment]
Okay it works now, thank you for your service.

**Comment by notpeter** [comment]
Glad to hear it! What ended up being the root cause?
I'd love to try and improve troubleshooting docs and want to know what to prioritize.
Thanks
